link-grammar/fast-match.c

```diff
@@ -203,6 +204,59 @@ static Match_node * add_to_left_table_list(Match_node * m, Match_node * l)
 }
 
 /**
+ * Validate that the uppercase part of two connectors are the same.
+ * Return true if so, false otherwise.
+ * FIXME: Use connector enumeration.
+ */
+static bool con_uc_eq(const Connector *c1, const Connector *c2)
+{
+
+	if (string_set_cmp(c1->string, c2->string)) return true;
+	if (c1->hash != c2->hash) return false;
+	if (c1->uc_length != c1->uc_length) return false;
+
+	const char *uc1 = &c1->string[c1->uc_start];
+	const char *uc2 = &c2->string[c2->uc_start];
+	/* We arrive here for less than 50% of the cases for "en" and
+	 * less then 20% of the cases for "ru", and the following
+	 * condition is always true.
+	 * How come 2 different uc strings never have the same hash value? */
+	if (0 == strncmp(uc1, uc2, c1->uc_length)) return true;
+
+	return false;
+}
+
+static Match_node **get_match_table_entry(unsigned int size, Match_node **t,
+                                          Connector * c, int dir)
+{
+		unsigned int h, s;
+
+		s = h = connector_hash(c) & (size-1);
+
+		if (dir == 1) {
+			while (NULL != t[h])
+			{
+				if (con_uc_eq(t[h]->d->right, c)) break;
+				h = (h + 1) & (size-1);
```

Comment:
this is  interesting and unusual.  Confused me at first; I see that it works, and that its needed.  Could you add a note that says "every hash bucket MUST have a unique upper-case connector, since easy_match_list only compares the lower-case parts".